File: src/main/java/minic/ui/workbench/MiniCStageListFactory.java
package minic.ui;

import minic.uiapi.UiCurrentStateDto;
import minic.uiapi.UiGlobalDataDto;
import minic.uiapi.UiStageDataDto;

import java.util.ArrayList;
import java.util.List;

/**
 * 根据 UI API DTO 生成侧边栏阶段列表展示数据。
 */
public final class MiniCStageListFactory {
    private static final List<StageInfo> STAGES = List.of(
            new StageInfo("source", "源码"),
            new StageInfo("preprocess", "预编译"),
            new StageInfo("lexer", "词法分析"),
            new StageInfo("parser", "语法分析"),
            new StageInfo("semantic", "语义分析"),
            new StageInfo("ir", "IR 降级"),
            new StageInfo("codegen", "代码生成"),
            new StageInfo("toolchain", "工具链"),
            new StageInfo("execution", "执行")
    );

    /**
     * 生成阶段列表。
     *
     * @param currentState 当前状态
     * @param currentStageData 当前阶段数据
     * @param globalData 全局数据
     * @return 阶段展示数据
     */
    public List<MiniCStageView> create(
            UiCurrentStateDto currentState,
            UiStageDataDto currentStageData,
            UiGlobalDataDto globalData
    ) {
        List<MiniCStageView> views = new ArrayList<>();
        String currentStage = currentState == null ? "lexer" : currentState.currentStage();
        int currentIndex = stageIndex(currentStage);
        for (int index = 0; index < STAGES.size(); index++) {
            StageInfo stage = STAGES.get(index);
            boolean active = stage.id().equals(currentStage);
            boolean done = index < currentIndex || active && currentStageData != null && currentStageData.completed();
            String state = state(active, done, globalData);
            int progress = progress(stage.id(), active, done, currentStageData);
            String detail = detail(stage.id(), active, currentStageData, globalData);
            views.add(new MiniCStageView(stage.id(), stage.title(), state, detail, progress));
        }
        return List.copyOf(views);
    }

    private int stageIndex(String id) {
        for (int index = 0; index < STAGES.size(); index++) {
            if (STAGES.get(index).id().equals(id)) {
                return index;
            }
        }
        return 0;
    }

    private String state(boolean active, boolean done, UiGlobalDataDto globalData) {
        if (hasErrors(globalData)) {
            return active ? "error" : done ? "done" : "queued";
        }
        if (active) {
            return "running";
        }
        if (done) {
            return "done";
        }
        return "queued";
    }

    private boolean hasErrors(UiGlobalDataDto globalData) {
        return globalData != null
                && globalData.diagnostics().stream().anyMatch(diagnostic -> "ERROR".equals(diagnostic.severity()));
    }

    private int progress(String stage, boolean active, boolean done, UiStageDataDto currentStageData) {
        if (done) {
            return 100;
        }
        if (!active || currentStageData == null || currentStageData.totalSteps() <= 0) {
            return 0;
        }
        long completed = currentStageData.completedSteps();
        long total = currentStageData.totalSteps();
        return (int) Math.max(0, Math.min(100, completed * 100 / total));
    }

    private String detail(
            String stage,
            boolean active,
            UiStageDataDto currentStageData,
            UiGlobalDataDto globalData
    ) {
        if (active && currentStageData != null) {
            return currentStageData.completedSteps() + " / " + currentStageData.totalSteps() + " · 当前阶段";
        }
        if (globalData == null) {
            return "等待会话启动";
        }
        return switch (stage) {
            case "source" -> "源码已加载";
            case "preprocess" -> globalData.preprocessSummary().isEmpty() ? "等待预编译" : "预处理产物已生成";
            case "lexer" -> globalData.tokenSummary().size() + " 个 token";
            case "parser" -> globalData.astSummary().size() + " 个 AST 项";
            case "semantic" -> globalData.semanticSummary().size() + " 个语义项";
            case "ir" -> globalData.irSummary().size() + " 个 IR 项";
            case "codegen" -> globalData.assemblySummary().size() + " 行汇编";
            case "toolchain" -> globalData.artifactSummary().isEmpty() ? "尚未生成产物" : "产物已就绪";
            case "execution" -> globalData.executionOutputSummary().isEmpty() ? "等待输入" : "运行完成";
            default -> "排队中";
        };
    }

    private record StageInfo(String id, String title) {
    }
}

File: uiweb/src/workbench/MiniCStageListFactory.ts
import type { JavaMirrorFile } from "../translation/javaMirror";
import type { UiCurrentStateDto, UiGlobalDataDto, UiStageDataDto } from "../translation/uiapi";
import { createMiniCStageView, type MiniCStageState, type MiniCStageViewModel } from "./MiniCStageView";

export const miniCStageListFactoryMirror = {
  "javaPath": "src/main/java/minic/uilocal/workbench/MiniCStageListFactory.java",
  "webPath": "uiweb/src/workbench/MiniCStageListFactory.ts",
  "packageName": "minic.uilocal",
  "exportName": "MiniCStageListFactory",
  "kind": "class",
  "imports": [
    "java.util.ArrayList",
    "java.util.List",
    "minic.uiapi.UiCurrentStateDto",
    "minic.uiapi.UiGlobalDataDto",
    "minic.uiapi.UiStageDataDto"
  ],
  "fields": [
    {
      "name": "STAGES",
      "signature": "private static final List<StageInfo>STAGES="
    }
  ],
  "methods": [
    {
      "name": "create",
      "signature": "create(UiCurrentStateDto currentState,UiStageDataDto currentStageData,UiGlobalDataDto globalData)"
    },
    {
      "name": "detail",
      "signature": "detail(String stage,boolean active,UiStageDataDto currentStageData,UiGlobalDataDto globalData)"
    },
    {
      "name": "hasErrors",
      "signature": "hasErrors(UiGlobalDataDto globalData)"
    },
    {
      "name": "progress",
      "signature": "progress(String stage,boolean active,boolean done,UiCurrentStateDto currentState,UiStageDataDto currentStageData)"
    },
    {
      "name": "stageIndex",
      "signature": "stageIndex(String id)"
    },
    {
      "name": "StageInfo",
      "signature": "StageInfo(String id,String title)"
    },
    {
      "name": "state",
      "signature": "state(boolean active,boolean done,UiGlobalDataDto globalData)"
    }
  ]
} as const satisfies JavaMirrorFile;

export class MiniCStageListFactory {
  static readonly mirror = miniCStageListFactoryMirror;

  readonly mirror = miniCStageListFactoryMirror;

  create(
    currentState: UiCurrentStateDto | null,
    currentStageData: UiStageDataDto | null,
    globalData: UiGlobalDataDto | null,
  ): readonly MiniCStageViewModel[] {
    const currentStage = currentState === null ? "source" : currentState.currentStage;
    const currentIndex = stageIndex(currentStage);
    return STAGES.map((stage, index) => {
      const active = stage.id === currentStage;
      const done =
        currentState !== null &&
        (index < currentIndex || (active && currentStageData !== null && currentStageData.completed));
      return createMiniCStageView(
        stage.id,
        stage.title,
        this.state(active, done, globalData),
        this.detail(stage.id, active, currentStageData, globalData),
        this.progress(stage.id, active, done, currentState, currentStageData),
      );
    });
  }

  state(active: boolean, done: boolean, globalData: UiGlobalDataDto | null): MiniCStageState {
    if (this.hasErrors(globalData)) {
      return active ? "error" : done ? "done" : "queued";
    }
    if (active) {
      return "running";
    }
    if (done) {
      return "done";
    }
    return "queued";
  }

  hasErrors(globalData: UiGlobalDataDto | null): boolean {
    return globalData !== null && globalData.diagnostics.some((diagnostic) => diagnostic.severity === "ERROR");
  }

  progress(
    stage: string,
    active: boolean,
    done: boolean,
    currentState: UiCurrentStateDto | null,
    currentStageData: UiStageDataDto | null,
  ): number {
    if (done) {
      return 100;
    }
    if (stage === "source" && active && currentState !== null) {
      return 100;
    }
    if (!active || currentStageData === null || currentStageData.totalSteps <= 0) {
      return 0;
    }
    return Math.max(
      0,
      Math.min(100, Math.trunc((currentStageData.completedSteps * 100) / currentStageData.totalSteps)),
    );
  }

  detail(
    stage: string,
    active: boolean,
    currentStageData: UiStageDataDto | null,
    globalData: UiGlobalDataDto | null,
  ): string {
    if (active && currentStageData !== null) {
      return `${currentStageData.completedSteps} / ${currentStageData.totalSteps} · 当前阶段`;
    }
    if (globalData === null) {
      return "等待会话启动";
    }
    switch (stage) {
      case "source":
        return "源码已加载";
      case "preprocess":
        return globalData.preprocessSummary.length === 0 ? "等待预编译" : "预处理产物已生成";
      case "lexer":
        return `${globalData.tokenSummary.length} 个 token`;
      case "parser":
        return `${globalData.astSummary.length} 个 AST 项`;
      case "semantic":
        return `${globalData.semanticSummary.length} 个语义项`;
      case "ir":
        return `${globalData.irSummary.length} 个 IR 项`;
      case "codegen":
        return `${globalData.assemblySummary.length} 行汇编`;
      case "toolchain":
        return globalData.artifactSummary.length === 0 ? "尚未生成产物" : "产物已就绪";
      case "execution":
        return globalData.executionOutputSummary.length === 0 ? "等待输入" : "运行完成";
      default:
        return "排队中";
    }
  }

  summary(): string {
    return `MiniCStageListFactory: ${this.mirror.methods.length} methods, ${this.mirror.fields.length} fields`;
  }
}

interface StageInfo {
  readonly id: string;
  readonly title: string;
}

const STAGES: readonly StageInfo[] = [
  { id: "source", title: "源码" },
  { id: "preprocess", title: "预编译" },
  { id: "lexer", title: "词法分析" },
  { id: "parser", title: "语法分析" },
  { id: "semantic", title: "语义分析" },
  { id: "ir", title: "IR 降级" },
  { id: "codegen", title: "代码生成" },
  { id: "toolchain", title: "工具链" },
  { id: "execution", title: "执行" },
];

function stageIndex(id: string): number {
  const index = STAGES.findIndex((stage) => stage.id === id);
  return index < 0 ? 0 : index;
}

export default MiniCStageListFactory;
